add pause and resume methods to storing process.

odedhutzler opened the pull request:
I added resume and pause methods to the Storage object.

We wanted to add the ability for the user / application to pause a download and resume it (e.g. upon closing the window).
Pausing a download is done on the local storage object. Resuming a download is done with the offline manifest URI (i assumed it will be saved by the application and not by Shaka).

The pause method signals the download manager and adds `downloadStatus = paused` to the manifest object. The `downloadSegment_` function was changed to support this flag. After pause is triggered, it will populate the manifest with the segment with the `dataKey` set to ` null` (and ` startByte, endByte, position`).

The resume method retrieves the manifest from the indexdb. 
It adds all the downloaded segments (those with `dataKey != null`) to the new manifestDB object.
All the pending segments are added to the `segments_` object in the `download_manager`.

I also changed the demo offline section to demonstrate the feature ability.

I still haven't changed and added tests - i will add them if you think i'm on the right track in implementing this methods.

odedhutzler commented:
I signed it!

ismena commented on demo/index.html:
```diff
@@ -257,6 +257,10 @@ <h1>Shaka Player <span id="version"></span></h1>
         <div>
           <button id="storeDeleteButton">Store</button>
           <span id="storeDeleteHelpText"></span>
+          <button id="pauseBtn">Pause</button>
```
Please change the names to pauseDownloadButton/resumeDownloadButton or similar (I assume that's what they do, right?). pause/resume button sounds like it's pausing/resuming the video.

odedhutzler replied:
Yes :)

ismena commented on demo/offline_section.js:
```diff
@@ -219,10 +225,26 @@ shakaDemo.storeDeleteAsset_ = function() {
 };
 
+/**
+ * @return {void}
```
You don't have to specify return type if it's void.

ismena commented on lib/media/segment_reference.js:
```diff
@@ -47,6 +47,9 @@ shaka.media.InitSegmentReference = function(uris, startByte, endByte) {
   this.endByte = endByte;
+
+  /** @const {!Array.<string>} */
+  this.uri;
```
1) rename to this.uris
2) should the variable hold anything at this point? If it doesn't have a value yet, use this.uris = [].

ismena commented on lib/media/segment_reference.js:
```diff
@@ -137,6 +140,9 @@ shaka.media.SegmentReference = function(
   this.endByte = endByte;
+
+  /** @const {?number} */
+  this.dataKey;
```
Same. If there is no value yet, initialize to null.

ismena commented on lib/offline/db_engine.js:
```diff
@@ -229,7 +229,7 @@ shaka.offline.DBEngine.prototype.removeSegments =
   return this.remove_(
       shaka.offline.DBEngine.Store.SEGMENT,
-      keys,
+      keys.filter(function(key) { return key !== null; }),
```
We typically prefer '==/!=' to '!==/===' unless there is a good reason to use the latter.

(Here and below)

odedhutzler replied:
@ismena, changed. But why do you prefer == over ===?

ismena commented on lib/offline/download_manager.js:
```diff
@@ -74,6 +74,9 @@ shaka.offline.DownloadManager = function(
   this.progressListeners_ = [];
+
+  /** @public {?number} */
+  this.offlineManifestStorageId;
```
Same. Please initialize to null if no value yet.

ismena commented on demo/offline_section.js:
```diff
@@ -149,6 +152,8 @@ shakaDemo.setupOfflineAssets_ = function() {
 };
 
+var storage; // eslint-disable-line no-var
+var savedOfflineUri; // eslint-disable-line no-var
```
use 'let' instead of 'var'

ismena commented:
@TheModMaker could you please review the logic

TheModMaker commented:
First I don't think we can accept this at the moment.  We want to make some major changes to the Storage internals as part of the fix to #1248.  So I think we should wait to work on this feature until that is fixed and the changes are made.  Especially since this will involve changes to the database contents, which will need reverse-compatibility.

Since this is a big feature, I would suggest discussing the design in #1301 first so we can all agree on the correct way to do this.  For example, I don't like storing every segment urls even if we aren't using pause/resume.

joeyparrish commented:
This also connects with #879, in which this pause/resume concept will be handled by background fetch.  I'm not against giving apps an explicit API to pause/resume, but the ultimate design needs to consider all of this.

odedhutzler commented:
@ismena , @TheModMaker & @joeyparrish - thanks for reviewing.

@ismena I modified according to your comments.
@TheModMaker I removed the segments urls from the segments object if already downloaded - definitely redundant data :). 

Regarding the big changes to the DB & storage internals (and background fetch) - do you have anything in mind already? some kind of design? i would be happy to follow it and change my PR accordingly.

joeyparrish commented:
I'm sorry, but we will need to put this on hold until we close #1248.  I also apologize that we don't have a document showing the plan for that.

Thank you for contributing, though.  If you are still interested in implementing this after we finish transforming the internals of our storage system, we would be happy to discuss a design in #1301.  Thanks!

odedhutzler commented:
@joeyparrish no worries, its understandable. I would be happy to be part of the design & implementation.

Thanks.

avelad commented:
Can you rebase and update the code? Thanks!

avelad commented:
Closing due to inactivity.
